### ISSUE TITLE
about group loss

### ISSUE THREAD

**Issue by hhaAndroid** [cswluo/SEF#1]
Hello, thank you for your work. 
After reading the paper and code, I found that Loss in the semantic grouping part is very difficult to understand, and the writing in the paper is relatively simple. Can you explain briefly？

LocalMaxGlobalMin

`    
   def forward(self, x):
        x = x.pow(2)  
        intra_x = []
        inter_x = []
        for i in range(self.nparts):
            if i == 0:
                intra_x.append((1 - x[:, :self.seps[i], :self.seps[i]]).mean())
            else:
                intra_x.append((1 - x[:, self.seps[i - 1]:self.seps[i], self.seps[i - 1]:self.seps[i]]).mean())
                inter_x.append(x[:, self.seps[i - 1]:self.seps[i], :self.seps[i - 1]].mean())

        loss = self.rho * 0.5 * (sum(intra_x) / self.nparts + sum(inter_x) / (self.nparts * (self.nparts - 1) / 2))

        return loss
 `

**Comment by cswluo** [comment]
The input variable, x, is a 3d tensor with the dimensionality nsamples * nfeatures * nfeatures.  For each slice (nfeature * nfeature), it should be considered as a block matrix (also it is a symmetric matrix). The blocks on the main diagonal are correlations of feature channels from the same group while the off-diagonal blocks are those from different groups. 

self.seps stores the indices of boundaries between groups. The equation provides the average correlation by dividing the total correlation of each type by the number of corresponding blocks.

hope it's helpful to you.

**Comment by cswluo** [comment]
> Hello, thank you for your work.
> After reading the paper and code, I found that Loss in the semantic grouping part is very difficult to understand, and the writing in the paper is relatively simple. Can you explain briefly？
> 
> LocalMaxGlobalMin
> 
> `
> def forward(self, x):
> x = x.pow(2)
> intra_x = []
> inter_x = []
> for i in range(self.nparts):
> if i == 0:
> intra_x.append((1 - x[:, :self.seps[i], :self.seps[i]]).mean())
> else:
> intra_x.append((1 - x[:, self.seps[i - 1]:self.seps[i], self.seps[i - 1]:self.seps[i]]).mean())
> inter_x.append(x[:, self.seps[i - 1]:self.seps[i], :self.seps[i - 1]].mean())
> 
> ```
>     loss = self.rho * 0.5 * (sum(intra_x) / self.nparts + sum(inter_x) / (self.nparts * (self.nparts - 1) / 2))
> 
>     return loss
> ```
> 
> `

The input variable, x, is a 3d tensor with the dimensionality nsamples * nfeatures * nfeatures. For each slice (nfeature * nfeature), it should be considered as a block matrix (also it is a symmetric matrix). The blocks on the main diagonal are correlations of feature channels from the same group while the off-diagonal blocks are those from different groups.

self.seps stores the indices of boundaries between groups. The equation provides the average correlation by dividing the total correlation of each type by the number of corresponding blocks.

hope it's helpful to you.
